packages/core/src/validate/run.ts

```diff
@@ -562,17 +572,25 @@ function* getLinkingErrors(
 }
 
 function* getInternalFunctionStorageErrors(
-  contractDef: ContractDefinition,
+  contractOrStructDef: ContractDefinition | StructDefinition,
+  deref: ASTDereferencer,
   decodeSrc: SrcDecoder,
 ): Generator<ValidationError> {
   // Note: Solidity does not allow annotations for non-public state variables, so this cannot be skipped with annotations
-  for (const variableDec of findAll('VariableDeclaration', contractDef)) {
+  for (const variableDec of findAll('VariableDeclaration', contractOrStructDef)) {
     if (variableDec.typeName?.nodeType === 'FunctionTypeName' && variableDec.typeName.visibility === 'internal') {
+      // Find internal function types directly in this node's scope
       yield {
         kind: 'internal-function-storage',
         name: variableDec.name,
         src: decodeSrc(variableDec),
       };
+    } else if (variableDec.typeName?.nodeType === 'UserDefinedTypeName') {
```

Comment:
If a variable is a mapping we would want to inspect the contents of the mapping value. The same for arrays and so on. I think we need a recursive traversal through the type.